plugin/pkg/scheduler/extender.go

```diff
@@ -82,53 +84,80 @@ func NewHTTPExtender(config *schedulerapi.ExtenderConfig, apiVersion string) (al
 		Timeout:   config.HTTPTimeout,
 	}
 	return &HTTPExtender{
-		extenderURL:    config.URLPrefix,
-		apiVersion:     apiVersion,
-		filterVerb:     config.FilterVerb,
-		prioritizeVerb: config.PrioritizeVerb,
-		weight:         config.Weight,
-		client:         client,
+		extenderURL:      config.URLPrefix,
+		apiVersion:       apiVersion,
+		filterVerb:       config.FilterVerb,
+		prioritizeVerb:   config.PrioritizeVerb,
+		weight:           config.Weight,
+		client:           client,
+		nodeCacheCapable: config.NodeCacheCapable,
 	}, nil
 }
 
 // Filter based on extender implemented predicate functions. The filtered list is
 // expected to be a subset of the supplied list. failedNodesMap optionally contains
 // the list of failed nodes and failure reasons.
-func (h *HTTPExtender) Filter(pod *v1.Pod, nodes []*v1.Node) ([]*v1.Node, schedulerapi.FailedNodesMap, error) {
-	var result schedulerapi.ExtenderFilterResult
+func (h *HTTPExtender) Filter(pod *v1.Pod, nodes []*v1.Node, nodeNameToInfo map[string]*schedulercache.NodeInfo) ([]*v1.Node, schedulerapi.FailedNodesMap, *schedulerapi.Annotations, error) {
+	var (
+		result     schedulerapi.ExtenderFilterResult
+		nodeItems  []v1.Node
+		nodeNames  *[]string
+		nodeResult []*v1.Node
+	)
 
 	if h.filterVerb == "" {
-		return nodes, schedulerapi.FailedNodesMap{}, nil
+		return nodes, schedulerapi.FailedNodesMap{}, nil, nil
 	}
 
-	nodeItems := make([]v1.Node, 0, len(nodes))
-	for _, node := range nodes {
-		nodeItems = append(nodeItems, *node)
+	if h.nodeCacheCapable {
+		nodeNameSlice := make([]string, 0, len(nodes))
+		for _, node := range nodes {
+			nodeNameSlice = append(nodeNameSlice, node.Name)
+		}
+		nodeNames = &nodeNameSlice
+	} else {
+		nodeItems = make([]v1.Node, 0, len(nodes))
+		for _, node := range nodes {
+			nodeItems = append(nodeItems, *node)
+		}
 	}
+
 	args := schedulerapi.ExtenderArgs{
-		Pod:   *pod,
-		Nodes: v1.NodeList{Items: nodeItems},
+		Pod:       *pod,
+		Nodes:     &v1.NodeList{Items: nodeItems},
+		NodeNames: nodeNames,
 	}
 
 	if err := h.send(h.filterVerb, &args, &result); err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 	if result.Error != "" {
-		return nil, nil, fmt.Errorf(result.Error)
+		return nil, nil, nil, fmt.Errorf(result.Error)
 	}
 
-	nodeResult := make([]*v1.Node, 0, len(result.Nodes.Items))
-	for i := range result.Nodes.Items {
-		nodeResult = append(nodeResult, &result.Nodes.Items[i])
+	if h.nodeCacheCapable {
+		nodeResult = make([]*v1.Node, 0, len(*result.NodeNames))
```

Comment:
Please add a defensive check for misbehaving extender. result.NodeNames can be nil.